### ISSUE TITLE
[Clang-repl] Assertion "Narrow integer argument must have a valid extension type." failed.

### ISSUE THREAD

**Issue by JonPsson1** [llvm/llvm-project#109658]
With the recent strengthening of the handling of extensions of narrow integer arguments (commit 1412022), it is now possible to detect cases of missing argument extensions, although there are still some issues that keeps this disabled by default. Many tests under Interpreter/ fails, for example:

```
cat ~/llvm-project/clang/test/Interpreter/assigment-with-implicit-ctor.cpp | ~/llvm-project/build/bin/clang-repl -argext-abi-check
good
SystemZISelLowering.cpp:9861: void llvm::SystemZTargetLowering::verifyNarrowIntegerArgs(const llvm::SmallVectorImpl<llvm::ISD::OutputArg>&, bool) const: Assertion `(VT != MVT::i32 || (Flags.isSExt() || Flags.isZExt() || Flags.isNoExt())) && "Narrow integer argument must have a valid extension type."' failed.
```

What this means is that there is a narrow (<=32 bit) integer argument that is neither sign or zero extended, and is also not marked as noext, which should be done for "struct-in-reg". So this may or may not be a "bug" (missing sign/zero extension), or it could be an argument that should be marked as noext. See https://llvm.org/docs/LangRef.html#parameter-attributes.

CCing some folks with the hope of help with this, thanks: @vgvassilev @hahnjo @vitalybuka @nico @weliveindetail @nikic

**Comment by JonPsson1** [comment]
With https://github.com/llvm/llvm-project/pull/109699, I get
```
cat ~/llvm-project/clang/test/Interpreter/assigment-with-implicit-ctor.cpp | ~/llvm-project/build/bin/clang-repl -argext-abi-check
good
ERROR: Missing extension attribute of returned value from function:
i32 @atexit(ptr)

```
My guess is that the returned value from @atexit needs sign or zero extension...

CCing some folks with the hope of help with this, thanks: @vgvassilev @hahnjo @vitalybuka @nico @weliveindetail @nikic

**Comment by vgvassilev** [comment]
Thanks for submitting a bug report and debugging it too. I am not sure what is going on here I think your guess is probably the best clue I have, too...

**Comment by nikic** [comment]
Presumably the problem is https://github.com/llvm/llvm-project/blob/cc7b24a4d125e9a81480aaaa961a2b963bbb2ea2/clang/lib/CodeGen/CGDeclCXX.cpp#L358-L359 Which creates the function without applying necessary ABI attributes.

I'd hope that clang codegen has *some* method to generate these kinds of runtime functions with appropriate attributes without requiring special casing for each one, but I wouldn't know what it is. Maybe the clang codegen maintainers (@rjmccall @efriedma-quic) have some pointers.

**Comment by JonPsson1** [comment]
My understanding is that you really have to do this specifically for each one as you have to specify if the integer is signed or unsigned, which there is otherwise no way to know. 

There is the `TLI->getExtAttrForI32Param(Triple, bool Signed)`, which seems to be used also outside BuildLibCalls, e.g. by the AddressSanitizer. This is however merely the next step after you know the signedness of the argument to let the target further specify what (if any) extension attribute should actually be added. Not sure however if that is required outside of libcalls, though, but it seems it would be reasonable.

**Comment by efriedma-quic** [comment]
CodeGenModule::CreateRuntimeFunction doesn't have enough information to mark up functions correctly, given its current signature.  I don't think we currently have an alternative.  We could add one pretty easily, though; we just a version that takes a clang::FunctionType instead of an llvm::FunctionType.

**Comment by rjmccall** [comment]
Most runtime functions are specified in terms of the C ABI, so yeah, we should probably change up how we emit runtime calls in general to go through the standard call-emission functions.  Maybe we can preserve a fast path for common runtime calls by designing a system for caching the runtime function and let just it also record when all the arguments use simple direct argument passing without any attributes.

**Comment by JonPsson1** [comment]
The improvement to dump the function signature for cases of missing extension attributes has now been committed.

**Comment by vgvassilev** [comment]
Can you link the commit here for future reference?

**Comment by JonPsson1** [comment]
> Can you link the commit here for future reference?

f9fbfc5

**Comment by JonPsson1** [comment]
ping...

**Comment by nikic** [comment]
@JonPsson1 Anything in particular you are waiting for here? @efriedma-quic's suggested approach to fixing this sounds reasonable to me.

**Comment by JonPsson1** [comment]
> @JonPsson1 Anything in particular you are waiting for here? @efriedma-quic's suggested approach to fixing this sounds reasonable to me.

Ah, well... I thought maybe someone who was actively working on Clang-repl might have a look at this. If not, I suppose I could give it a try :)

**Comment by vgvassilev** [comment]
I am not entirely sure what the problem is. Probably would be best to give it a try.

**Comment by nikic** [comment]
> > @JonPsson1 Anything in particular you are waiting for here? @efriedma-quic's suggested approach to fixing this sounds reasonable to me.
> 
> Ah, well... I thought maybe someone who was actively working on Clang-repl might have a look at this. If not, I suppose I could give it a try :)

I don't think the problem here is really related to clang-repl. You can probably reproduce this with plain clang, `-fregister-global-dtors-with-atexit` and a global dtor.

Though the overall problem of runtime functions is even more general than that, e.g. I think that https://github.com/llvm/llvm-project/issues/109654 is really the same core problem, it's also a function created with CreateRuntimeFunction. So once the basic infrastructure for this is in place, variants of this issue can be fixed by migrating to the new API.

**Comment by JonPsson1** [comment]
I tried fixing this the same way as I have (at least initially) done in #109654 , but that attribute somehow did not make it to the CodeGen phase..(?). However, "atexit" is "hidden", so this is actually not necessary for correctness. So instead, extend SystemZTargetLowering::isFullyInternal() to not only recognize local linkage, but also private linkage and hidden visibility as exclusion criterias:
```

diff --git a/llvm/lib/Target/SystemZ/SystemZISelLowering.cpp b/llvm/lib/Target/SystemZ/SystemZISelLowering.cpp
index abf414fd1ffb..7f4113317024 100644
--- a/llvm/lib/Target/SystemZ/SystemZISelLowering.cpp
+++ b/llvm/lib/Target/SystemZ/SystemZISelLowering.cpp
@@ -9815,11 +9815,12 @@ SDValue SystemZTargetLowering::lowerVECREDUCE_ADD(SDValue Op,
       DAG.getConstant(OpVT.getVectorNumElements() - 1, DL, MVT::i32));
 }
 
-// Only consider a function fully internal as long as it has local linkage
-// and is not used in any other way than acting as the called function at
-// call sites.
+// Only consider a function fully internal as long as it has local or private
+// linkage or is hidden and is not used in any other way than acting as the
+// called function at call sites.
 bool SystemZTargetLowering::isFullyInternal(const Function *Fn) const {
-  if (!Fn->hasLocalLinkage())
+  if (!Fn->hasLocalLinkage() && !Fn->hasHiddenVisibility() &&
+      !Fn->hasPrivateLinkage())
     return false;
   for (const User *U : Fn->users()) {
     if (auto *CB = dyn_cast<CallBase>(U)) {

```

IIUC hidden / private should also be legitimate "module local" flags, right?  @uweigand LGTY?

**Comment by nikic** [comment]
`hidden` refers to shared object visibility, see https://llvm.org/docs/LangRef.html#visibility-styles. This is orthogonal to the linkage and not relevant in this context.

**Comment by JonPsson1** [comment]
Hmm, I think that section is quite misleading then: "All Global Variables and Functions have one of the following visibility styles: ... “hidden” - Hidden style ... "the symbol will not be placed into the dynamic symbol table, so no other module (executable or shared library) can reference it directly.". 

So the atexit function is marked "hidden". Doesn't this mean that it is not visible to any other module?

```
(gdb) p F->dump()
define hidden i32 @atexit(ptr %0) {
entry:
  %1 = call i32 @__lljit.atexit_helper(ptr @__lljit.platform_support_instance, ptr @__dso_handle, ptr %0)
  ret i32 %1
}

$15 = void
(gdb) p F->hasLocalLinkage()
$16 = false
(gdb) p F->hasPrivateLinkage()
$17 = false
(gdb) p F->hasHiddenVisibility()
$18 = true

```

**Comment by nikic** [comment]
"Module" here is not an LLVM IR module, but a module in the sense of the parenthetical ("executable or shared library").

**Comment by nikic** [comment]
By the way, it looks like the verification currently fails to verify return values on calls. This passes verification but should not:
```
declare i32 @test(i32)

define signext i32 @test2() {
  %res = call i32 @test()
  ret i32 %res
}
```

(Checked this because your example has `atexit` as a definition, while I'd generally expect it to be a declaration only, with the definition provided by glibc -- I guess this might be the relation that clang-repl has to this.)

**Comment by JonPsson1** [comment]
That's interesting as I think this is actually not necessary to detect: whoever defines @test() needs to make sure the returned value is extended (which is checked for), but in this case this just means that @test2() needs to do the extension (which it does) as there is no attribute saying it was done by @test(). This is however a matter of optimization, not correctness, which has been the aim so far.

(I just checked and it seems clang normally emits the attribute for a function declaration, so this should typically be there.)

**Comment by lhames** [comment]
I think we probably just need to add the attribute to the `atexit` declaration that LLJIT injects. @JonPsson1 -- does the following look right to you?

```c++
diff --git a/llvm/lib/ExecutionEngine/Orc/LLJIT.cpp b/llvm/lib/ExecutionEngine/Orc/LLJIT.cpp
index d3dd3b6bedfb..d4b73c21f19f 100644
--- a/llvm/lib/ExecutionEngine/Orc/LLJIT.cpp
+++ b/llvm/lib/ExecutionEngine/Orc/LLJIT.cpp
@@ -7,6 +7,8 @@
 //===----------------------------------------------------------------------===//
 
 #include "llvm/ExecutionEngine/Orc/LLJIT.h"
+
+#include "llvm/Analysis/TargetLibraryInfo.h"
 #include "llvm/Config/llvm-config.h" // for LLVM_ENABLE_THREADS
 #include "llvm/ExecutionEngine/JITLink/EHFrameSupport.h"
 #include "llvm/ExecutionEngine/JITLink/JITLinkMemoryManager.h"
@@ -195,10 +197,14 @@ public:
     auto *IntTy = Type::getIntNTy(*Ctx, sizeof(int) * CHAR_BIT);
     auto *AtExitCallbackTy = FunctionType::get(VoidTy, {}, false);
     auto *AtExitCallbackPtrTy = PointerType::getUnqual(AtExitCallbackTy);
-    addHelperAndWrapper(*M, "atexit",
+    auto *AtExit = addHelperAndWrapper(*M, "atexit",
                         FunctionType::get(IntTy, {AtExitCallbackPtrTy}, false),
                         GlobalValue::HiddenVisibility, "__lljit.atexit_helper",
                         {PlatformInstanceDecl, DSOHandle});
+    Attribute::AttrKind AtExitExtAttr =
+      TargetLibraryInfo::getExtAttrForI32Return(J.getTargetTriple());
+    if (AtExitExtAttr != Attribute::None)
+      AtExit->addRetAttr(AtExitExtAttr);
 
     return J.addIRModule(JD, ThreadSafeModule(std::move(M), std::move(Ctx)));
   }
```

**Comment by nikic** [comment]
That looks right to me.

**Comment by JonPsson1** [comment]
@lhames @nikic yes, that looks right to me as well, and it fixes this issue. My only concern is if there were to be other similar cases here so that it would make sense to e.g. wrap the last 4 lines in utility function instead. But I guess maybe that could wait until we see that this is actually a recurring problem...

**Comment by lhames** [comment]
@JonPsson1 Yep. I've committed this as a one-off fix for now in 4245c00faf5e. We're going to revisit how `LLJIT` declares these functions in the not too distant future anyway -- that'll be a good time to consider a general solution.

**Comment by JonPsson1** [comment]
Thanks!
